stdlib/runtime/atof/decimal.gr

```diff
@@ -0,0 +1,621 @@
+// https://github.com/rust-lang/rust/blob/1cbc45942d5c0f6eb5d94e3b10762ba541958035/library/core/src/num/dec2flt/decimal.rs
+
+import WasmI32 from "runtime/unsafe/wasmi32"
+import WasmI64 from "runtime/unsafe/wasmi64"
+import Memory from "runtime/unsafe/memory"
+import { newInt32, allocateBytes } from "runtime/dataStructures"
+
+import {
+  _CHAR_CODE_PLUS,
+  _CHAR_CODE_MINUS,
+  _CHAR_CODE_0,
+  _CHAR_CODE_E,
+  _CHAR_CODE_e,
+  _CHAR_CODE_DOT,
+  is8Digits,
+} from "./common"
+
+primitive (&&): (Bool, Bool) -> Bool = "@and"
+primitive (||): (Bool, Bool) -> Bool = "@or"
+primitive (!): Bool -> Bool = "@not"
+
+export record Decimal {
+  /// The number of significant digits in the decimal.
+  mut numDigits: Int32,
+  /// The offset of the decimal point in the significant digits.
+  mut decimalPoint: Int32,
+  /// If the number of significant digits stored in the decimal is truncated.
+  mut truncated: Bool,
+  /// Buffer of the raw digits, in the range [0, 9].
+  digits: Bytes,
+}
+
+/// The maximum number of digits required to unambiguously round a float.
+///
+/// For a double-precision IEEE 754 float, this required 767 digits,
+/// so we store the max digits + 1.
+///
+/// We can exactly represent a float in radix `b` from radix 2 if
+/// `b` is divisible by 2. This function calculates the exact number of
+/// digits required to exactly represent that float.
+///
+/// According to the "Handbook of Floating Point Arithmetic",
+/// for IEEE754, with emin being the min exponent, p2 being the
+/// precision, and b being the radix, the number of digits follows as:
+///
+/// `−emin + p2 + ⌊(emin + 1) log(2, b) − log(1 − 2^(−p2), b)⌋`
+///
+/// For f32, this follows as:
+///     emin = -126
+///     p2 = 24
+///
+/// For f64, this follows as:
+///     emin = -1022
+///     p2 = 53
+///
+/// In Python:
+///     `-emin + p2 + math.floor((emin+ 1)*math.log(2, b)-math.log(1-2**(-p2), b))`
+@unsafe
+let _MAX_DIGITS = 768n
+/// The max digits that can be exactly represented in a 64-bit integer.
+@unsafe
+let _MAX_DIGITS_WITHOUT_OVERFLOW = 19n
+@unsafe
+export let _DECIMAL_POINT_RANGE = 2047n
+
+@unsafe
+let new = () => {
+  let digits = allocateBytes(_MAX_DIGITS)
+  Memory.fill(WasmI32.add(digits, 8n), 0n, _MAX_DIGITS)
+  let digits = WasmI32.toGrain(digits)
+  { numDigits: 0l, decimalPoint: 0l, truncated: false, digits }
+}
+
+/// Append a digit to the buffer.
+@unsafe
+export let tryAddDigit = (d, digit) => {
+  let (+) = WasmI32.add
+  let numDigits = WasmI32.load(WasmI32.fromGrain(d.numDigits), 8n)
+  if (WasmI32.ltU(numDigits, _MAX_DIGITS)) {
+    let digits = WasmI32.fromGrain(d.digits)
+    WasmI32.store8(digits + numDigits, digit, 8n)
+  }
+  d.numDigits = WasmI32.toGrain(newInt32(WasmI32.add(numDigits, 1n)))
+}
+
+/// Trim trailing zeros from the buffer.
+@unsafe
+let trim = d => {
+  let (!=) = WasmI32.ne
+  let (==) = WasmI32.eq
+  let (+) = WasmI32.add
+  let (-) = WasmI32.sub
+  // All of the following calls to `trim` are fine because:
```

Comment:
@jozanza Maybe this comment should refer more precisely to where the call site is (rather than "the following")?